### ISSUE TITLE
Notes app creates multiple notes when creating a new one

### ISSUE THREAD

**Issue by yostral** [stefan-niedermann/nextcloud-notes#1198]
Please use **[GitHub reactions](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/)** 👍 to show that you are affected by the same issue. Please don't comment if you have no relevant information to add!

**Describe the bug**

When I create a new note, and going back to the notes list, it has created multiple notes. They all have the same title, the one I gave, and the title is followed by (1), (2), (3), etc. Inside each note, there is the text I entered, but just the title in the first note, the first few words in the second note, a bit more words in the third, etc, until the last one with everything.

It's like if there was a incremental backup creating a new note each time, instead of backing up to the same note.

When I want to delete all the useless ones, so not the last one (which I have to rename), I have an error :

```
App Version: 3.4.1
App Version Code: 3004001
App Flavor: fdroid

Files App Version Code: 30150190

---

OS Version: 4.14.190-perf+(2104052320)
OS API Level: 30
Device: OnePlus7Pro
Manufacturer: OnePlus
Model (and Product): GM1910 (OnePlus7Pro_EEA)

---

com.nextcloud.android.sso.exceptions.NextcloudHttpRequestFailedException: La requête HTTP a échoué avec le statut HTTP :423
	at com.nextcloud.android.sso.api.AidlNetworkRequest.performNetworkRequestV2(AidlNetworkRequest.java:172)
	at com.nextcloud.android.sso.api.NextcloudAPI.performNetworkRequestV2(NextcloudAPI.java:171)
	at it.niedermann.owncloud.notes.persistence.SSOClient.requestFilesApp(SSOClient.java:27)
	at it.niedermann.owncloud.notes.persistence.NotesClient.requestServer(NotesClient.java:178)
	at it.niedermann.owncloud.notes.persistence.NotesClientV1.deleteNote(NotesClientV1.java:57)
	at it.niedermann.owncloud.notes.persistence.NotesServerSyncTask.pushLocalChanges(NotesServerSyncTask.java:126)
	at it.niedermann.owncloud.notes.persistence.NotesServerSyncTask.run(NotesServerSyncTask.java:72)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:462)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.lang.Thread.run(Thread.java:923)
Caused by: java.lang.IllegalStateException: {"errorType":"OCP\\Lock\\LockedException"}

	at com.nextcloud.android.sso.InputStreamBinder.processRequestV2(InputStreamBinder.java:429)
	at com.nextcloud.android.sso.InputStreamBinder.performNextcloudRequestAndBodyStreamV2(InputStreamBinder.java:125)
	at com.nextcloud.android.sso.InputStreamBinder.performNextcloudRequestV2(InputStreamBinder.java:108)
	at com.nextcloud.android.sso.aidl.IInputStreamService$Stub.onTransact(IInputStreamService.java:158)
	at android.os.Binder.execTransactInternal(Binder.java:1165)
	at android.os.Binder.execTransact(Binder.java:1134)

```

But not everytime... sometimes it deletes them without the need to restart the Notes app.

I can see every notes synchronize on my NC.

I have deleted the app data, re-signed in, no change, same behavior.

**To Reproduce**
1. Go to Notes Android app
2. Click on the '+' to create a new note
4. See the multiple notes and the error when you want to delete them.

**Expected behavior**

When creating a new note, it should create only one note.

**Screenshots**

https://nextcloud.yostral.me/s/jeStQ2grewZiXbr

**Smartphone (please complete the following information):**
 - Nextcloud Notes-Version (android app): 3.4.1
 - F-Droid or Play Store: F-droid
 - Android-Version: 11
 - Device: OnePlus 7 pro

**Server**
 - Nextcloud version: 21.0.1
 - Nextcloud Notes version (server app): 4.0.4

**Stacktrace**
Here is everything I have in my NC logs :

```
{"reqId":"tGUs7lwOp9hBW9HsQvhT","level":1,"time":"2021-05-04T22:35:21+00:00","remoteAddr":"xxx.xxx.xxx.xxx","user":"user1","app":"workflowengine","method":"POST","url":"/index.php/apps/notes/api/v1/notes?","message":"Flow handling done for event \\OCP\\Files::postCreate","userAgent":"Mozilla/5.0 (Android) Nextcloud-android/3.15.1","version":"21.0.1.1","id":"6091cc29582b6"}
```
```
{"reqId":"tGUs7lwOp9hBW9HsQvhT","level":1,"time":"2021-05-04T22:35:21+00:00","remoteAddr":"xxx.xxx.xxx.xxx","user":"user1","app":"workflowengine","method":"POST","url":"/index.php/apps/notes/api/v1/notes?","message":"Flow handling done for event \\OCP\\Files::postWrite","userAgent":"Mozilla/5.0 (Android) Nextcloud-android/3.15.1","version":"21.0.1.1","id":"6091cc2958282"}
```
```
{"reqId":"tGUs7lwOp9hBW9HsQvhT","level":1,"time":"2021-05-04T22:35:21+00:00","remoteAddr":"xxx.xxx.xxx.xxx","user":"user1","app":"no app in context","method":"POST","url":"/index.php/apps/notes/api/v1/notes?","message":"Deprecated event type for \\OCP\\Files::postWrite: Symfony\\Component\\EventDispatcher\\GenericEvent is used","userAgent":"Mozilla/5.0 (Android) Nextcloud-android/3.15.1","version":"21.0.1.1","id":"6091cc2958275"}
```

**Comment by korelstar** [comment]
The HTTP Status Code is "locked", i.e. the newly created file has been locked by someone. Therefore, the notes app can't write the content and on the next synchronization, it tries again (another empty note is created).

The server log indicates that a workflow (Nextcloud flow) was started. Maybe that workflow locked the file. Please try to deactivate all workflows on your Nextcloud.

**Comment by stefan-niedermann** [comment]
I think the root cause is that switching between edit and preview mode while creating a new note creates new notes on the server.

@yostral can you confirm that the behavior is correct if you do *not* switch between edit and preview mode until you have at least closed and reopened the note once?

This must have slipped while migrating from `SQLiteOpenHelper` to `Room`...

**Comment by yostral** [comment]
@korelstar : I deactivated the only flow I had, PDF conversion (not even working because no LibreOffice installed), but it is the same.

@stefan-niedermann : I don't switch from edit to preview mode. I create a note, so I'm in edit mode, and I directly go back to the list. If I create a note and close Notes without coming back, it's the same result after the next opening, multiple notes were created. If, when creating, I switch to view and then back to the list, it's the same as well.

I tried with another phone, a OnePlus 5 with Android 10, Android Notes 3.4.0, it's the same as well. So the problem may be on the server side? How do I find the last updated apps? Is it possible?

**Comment by korelstar** [comment]
I'm wondering why the Android is saying that the server sent a `LockedException` but you can't find a `LockedException` in the server log. Maybe you could search explicitly for that? It must be there!

In the past, there were some users that had problems with a `LockedException`. Most of them could fix it by deactivating a problemous app. A prominent example was the app ransomware detection, see e.g. https://github.com/nextcloud/notes/issues/468 and https://github.com/nextcloud/notes/issues/430 

Therefore, please try to temporarily deactivate other Nextcloud apps, especially non-official ones. Maybe, you have a similar problem.

**Comment by IngoBelka** [comment]
I can confirm as following

Writing a note, going back to the notes list, everything is fine.

Writing a note "name", switching between preview and edit, new notes "name (x)", where x is an increasing number, are created.

This only happens with new posts, not with existing ones.

**Comment by yostral** [comment]
(the https://github.com/nextcloud/notes/issues/430 was me :) )

So, I made few tests. I deactivated almost every app, all the 3rd party and many official ones. And the same thing happens. With absolutely _nothing_ in the logs. Neither in the journalctl of the server.

I tried with another NC I manage, still NC20.0.8, with "old" official apps only. And it is the same, and nothing in the logs as well.

So then I reinstalled the Android Notes app v3.3.1, from F-droid, and it works normally, without creating multiple notes.

In fact, it seems to happens only when I change the title of the note. If I write in the body, and then the first line is taken for the title, it doesn't seem to happen.

**Comment by korelstar** [comment]
@stefan-niedermann this sounds like the Android app is doing the requests in parallel. This should not be.

**Comment by stefan-niedermann** [comment]
Root cause of the issue seems to be, that the sync status is set to `LOCAL_EDITED` (where it should be `VOID`) when creating a new note.
This leads to pushing the same note to the server everytime when the automatic save or switching between preview / edit is triggered. Then the note gets synced back to the device and we have the same note multiple times.

I'll investigate this further...

**Comment by stefan-niedermann** [comment]
New updates: Before Room we only updated a few fields (excluding `remoteId`) in `updateNoteAndSync`. After Room, we just put in the whole `Note` entity (including `null` as `remoteId`).
I will make a workaround and re-read this field from the database in `updateNoteAndSync`, which *should*™ solve the issue. If it works, we can optimize this in the long term to
- either push the `remoteId` back to the `BaseNoteFragment`, so it won't pass `null` as `remoteId` when saving again **or**
- only update the fields which have changed as in the pre-Room implementation

... will keep you updated.

**Comment by stefan-niedermann** [comment]
The update is available on F-Droid now.

**Comment by IngoBelka** [comment]
I made the update right now and I can confirm, that in version 3.4.2 the bug is solved, as far as I can see.

**Comment by stefan-niedermann** [comment]
Thank you for the feedback 👍 

Closing this for now - feel free to open a new issue in case the problem reappears 🙂
